[Serve] [Docs] Async io + Ray Serve best practices (#58909)
ray-project/ray: doc/source/serve/advanced-guides/asyncio-best-practices.md

```

## Quick guide: choose between `async def` and `def`

Use this decision table as a starting point:

| Workload type | Recommended handler | Why |
| --- | --- | --- |
| I/O-bound (databases, HTTP calls, queues) | `async def` | Lets the event loop handle many requests while each waits on I/O. |
| CPU-bound (model inference, heavy numeric compute) | `def` or `async def` with offload | Async alone doesn't make CPU work faster. You need more replicas, threads, or native parallelism. |
| Streaming responses | `async def` generator | Integrates with backpressure and non-blocking iteration. |
| FastAPI ingress (`@serve.ingress`) | `def` or `async def` | FastAPI runs `def` endpoints in a threadpool, so they don't block the loop. |

You can always mix styles inside a deployment. For example, use `async def` for request handling and offload CPU-heavy parts to a threadpool.

## How Ray Serve executes your code

At a high level, requests go through a router to a replica actor that runs your code:

```text
Client
   ↓
Serve router (asyncio loop A)
   ↓
Replica actor
   ├─ System / control loop
   └─ User code loop (your handlers)
       └─ Optional threadpool for sync methods
```

> Key ideas:

Reviewer: ```suggestion
The following are the key ideas to consider when deciding to use `async def` or `def`:
```

- Serve uses asyncio event loops for routing and for running replicas.

> - By default, user code runs on a separate event loop from the replica's main/control loop, so blocking user code does not interfere with health checks and autoscaling.

Reviewer: ```suggestion
- By default, user code runs on a separate event loop from the replica's main/control loop, so blocking user code doesn't interfere with health checks and autoscaling.
```

> - Depending on configuration, `def` handlers may run directly on the user event loop (blocking) or in a threadpool (non-blocking for the loop).

Reviewer: ```suggestion
- Depending on the configuration, `def` handlers may run directly on the user event loop (blocking) or in a threadpool (non-blocking for the loop).
```

```{literalinclude} ../doc_code/asyncio_best_practices.py
:start-after: __echo_async_begin__
:end-before: __echo_async_end__
:language: python
```

- `async def __call__` runs directly on the replica's user event loop.
- While this handler awaits `asyncio.sleep`, the loop is free to start handling other requests.

For a synchronous deployment:

```{literalinclude} ../doc_code/asyncio_best_practices.py
:start-after: __blocking_echo_begin__
:end-before: __blocking_echo_end__
:language: python
```

How this method executes depends on configuration:

- With `RAY_SERVE_RUN_SYNC_IN_THREADPOOL=0` (current default), `__call__` runs directly on the user event loop and blocks it for 1 second.
- With `RAY_SERVE_RUN_SYNC_IN_THREADPOOL=1`, Serve offloads `__call__` to a threadpool so the event loop stays responsive.

### FastAPI ingress (`@serve.ingress`)

When you use FastAPI ingress, FastAPI controls how endpoints run:

```{literalinclude} ../doc_code/asyncio_best_practices.py
:start-after: __fastapi_deployment_begin__
:end-before: __fastapi_deployment_end__
:language: python
```

Important differences:

- In FastAPI, `def` endpoints are always dispatched to a threadpool.

> - In pure Serve, `def` methods currently run on the event loop unless you opt into threadpool behavior.

Reviewer: ```suggestion
- In pure Serve, `def` methods run on the event loop unless you opt into threadpool behavior.
```

> ## Blocking vs non-blocking in practice

Reviewer: ```suggestion
## Blocking versus non-blocking in practice
```

Blocking code keeps the event loop from processing other work. Non-blocking code yields control back to the loop when it is waiting on something.

### Blocking I/O vs asynchronous I/O

Blocking I/O example:

```{literalinclude} ../doc_code/asyncio_best_practices.py
:start-after: __blocking_http_begin__
:end-before: __blocking_http_end__
:language: python
```

Even though the method is `async def`, `requests.get` blocks the loop. No other requests can run on this replica during the request call. Blocking in `async def` is still blocking.

Non-blocking equivalent with async HTTP client:

```{literalinclude} ../doc_code/asyncio_best_practices.py
:start-after: __async_http_begin__
:end-before: __async_http_end__
:language: python
```

Non-blocking equivalent using a threadpool:

```{literalinclude} ../doc_code/asyncio_best_practices.py
:start-after: __threaded_http_begin__
:end-before: __threaded_http_end__
:language: python
```

## Concurrency does not equal parallelism in Python

> It is common to expect `async` code to "use all the cores" or make CPU-heavy code faster. That is not what asyncio does.

Reviewer: ```suggestion
It's common to expect `async` code to "use all the cores" or make CPU-heavy code faster. asyncio doesn't do that.
```

:language: python
```

- Up to 100 requests can be in-flight per replica.
- While one request is waiting, the event loop can work on others.

### Blocking `def` handlers and `max_ongoing_requests`

With a blocking `def` handler that runs on the event loop (threadpool disabled), `max_ongoing_requests` does not give you the concurrency you expect:

```{literalinclude} ../doc_code/asyncio_best_practices.py
:start-after: __blocking_cpu_begin__
:end-before: __blocking_cpu_end__
:language: python
```

In this case:

- The event loop can only run one handler at a time.
- Even though `max_ongoing_requests=100`, the replica effectively processes requests serially.

If you enable the sync-in-threadpool behavior (see the next section), each in-flight request can run in a thread:

```{literalinclude} ../doc_code/asyncio_best_practices.py
:start-after: __cpu_with_threadpool_begin__
:end-before: __cpu_with_threadpool_end__
:language: python
```

Now:

- Up to `max_ongoing_requests` calls can be running at once.
- Real throughput depends on:
  - How many threads the threadpool uses.
  - Whether your workload is CPU-bound or GIL-releasing.
  - Underlying native libraries and system resources.

> For heavily CPU-bound workloads, it is usually better to:

Reviewer: ```suggestion
For heavily CPU-bound workloads, it's usually better to:
```

- Keep `max_ongoing_requests` modest (to avoid queueing too many heavy tasks), and
- Scale **replicas** (`num_replicas`) rather than pushing a single replica's concurrency too high.

## Environment flags and sync-in-threadpool warning

Ray Serve exposes several environment variables that control how user code interacts with event loops and threads.

### `RAY_SERVE_RUN_SYNC_IN_THREADPOOL`

By default (`RAY_SERVE_RUN_SYNC_IN_THREADPOOL=0`), synchronous methods in a deployment run directly on the user event loop. To help you migrate to a safer model, Serve emits a warning like:

> `RAY_SERVE_RUN_SYNC_IN_THREADPOOL_WARNING`: Calling sync method '...' directly on the asyncio loop. In a future version, sync methods will be run in a threadpool by default...

This warning means:

- You have a `def` method that is currently running on the event loop.

> - In a future version, that method will run in a threadpool instead.

Reviewer: ```suggestion
- In a future version, that method runs in a threadpool instead.
```

### `RAY_SERVE_RUN_ROUTER_IN_SEPARATE_LOOP`

Serve's request router is also run on its own event loop by default:

```bash
export RAY_SERVE_RUN_ROUTER_IN_SEPARATE_LOOP=1  # default
```

This ensures:

- The router can continue routing and load balancing requests even if some replicas are running slow user code.

Disabling this:

```bash
export RAY_SERVE_RUN_ROUTER_IN_SEPARATE_LOOP=0
```

makes the router share an event loop with other work. This can reduce overhead in advanced, highly optimized scenarios, but makes the system more sensitive to blocking operations.

For most production deployments, you should keep the defaults (`1`) for both separate-loop flags.

## Batching and streaming semantics

Batching and streaming both rely on the event loop to stay responsive. They don't change where your code runs: batched handlers and streaming handlers still run on the same user event loop as any other handler. This means that if you add batching or streaming on top of blocking code, you can make event loop blocking effects much worse.

### Batching

> When you enable batching, Serve groups multiple incoming requests together and passes them to your handler as a list. The handler still runs on the user event loop, but each call now processes many requests at once instead of just one. If that batched work is blocking, it blocks the event loop for all of those requests at the same time.

Reviewer 2: May be good to call out the interplay with concurrent batch api - i.e. its only effective if there is actually i/o inside the batch execution code

```

## Quick guide: choose between `async def` and `def`

Use this decision table as a starting point:

| Workload type | Recommended handler | Why |
| --- | --- | --- |
| I/O-bound (databases, HTTP calls, queues) | `async def` | Lets the event loop handle many requests while each waits on I/O. |
| CPU-bound (model inference, heavy numeric compute) | `def` or `async def` + offload | Async alone does not make CPU work faster. You need more replicas, threads, or native parallelism. |
| Streaming responses | `async def` generator | Integrates with backpressure and non-blocking iteration. |
| FastAPI ingress (`@serve.ingress`) | `def` or `async def` | FastAPI runs `def` endpoints in a threadpool, so they do not block the loop. |

You can always mix styles inside a deployment. For example, use `async def` for request handling and offload CPU-heavy parts to a threadpool or Ray tasks.

## How Ray Serve executes your code

At a high level, requests go through a router to a replica actor that runs your code:

```text
Client
   ↓
Serve router (asyncio loop A)
   ↓
Replica actor
   ├─ System / control loop
   └─ User code loop (your handlers)
       └─ Optional threadpool for sync methods

> ```

Reviewer 3: really nice to have this diagram - sets a good mental model for readers.

One nitpick about Serve in general, is the term handle is a bit overloaded, a deploymentHandle will invoke a router, and then the request will get handled by my user code :) - my vote would have went for deploymentRouter as a name instead ...

I get that DeploymentHandle though is more semantically consistent with an ActorHandle in Ray core so 🤷

        # Runs directly on FastAPI's asyncio loop.
        await asyncio.sleep(1)
        return "ok"
```

Important differences:

- In FastAPI, `def` endpoints are always dispatched to a threadpool.
- In pure Serve, `def` methods currently run on the event loop unless you opt into threadpool behavior.

## Blocking vs non-blocking in practice

Blocking code keeps the event loop from processing other work. Non-blocking code yields control back to the loop when it is waiting on something.

### Blocking in `async def` is still blocking

This handler looks asynchronous but is not:

```python
@serve.deployment
class BadHandler:
    async def __call__(self, request):
        # ❌ Blocking: time.sleep blocks the entire event loop.
        import time
        time.sleep(2)
        return "done"
```

Even though the method is `async def`, `time.sleep` blocks the loop. No other requests can run on this replica during those 2 seconds.

The correct version uses `asyncio.sleep`:

```python
@serve.deployment
class GoodHandler:
    async def __call__(self, request):
        # ✅ Non-blocking: allows other requests to run.
        await asyncio.sleep(2)
        return "done"
```

> ### Blocking I/O vs asynchronous I/O

Reviewer 3: I think there is a bit of redundancy/verbosity here, I suggest perhaps we remove the BadHandler, GoodHandler examples given the examples BlockingHTTP, AsyncHTTP are more realistic and cover similar concepts of an async defined coroutine still being blocked due to interleaving synchronous code.

> `RAY_SERVE_RUN_SYNC_IN_THREADPOOL_WARNING`: Calling sync method '...' directly on the asyncio loop. In a future version, sync methods will be run in a threadpool by default...

This warning means:

- You have a `def` method that is currently running on the event loop.
- In a future version, that method will run in a threadpool instead.

You can opt in to the future behavior now by setting:

```bash
export RAY_SERVE_RUN_SYNC_IN_THREADPOOL=1
```

When this flag is `1`:

- Serve runs synchronous methods in a threadpool.
- The event loop is free to keep serving other requests while sync methods run.

Before enabling this in production, make sure:

- Your handler code and any shared state are thread-safe.
- Your model objects can safely be used from multiple threads, or you protect them with locks.

### `RAY_SERVE_RUN_USER_CODE_IN_SEPARATE_THREAD`

By default, Serve runs user code in a separate event loop from the replica's main/control loop:

```bash
export RAY_SERVE_RUN_USER_CODE_IN_SEPARATE_THREAD=1  # default
```

This isolation:

- Protects system tasks (health checks, controller communication) from being blocked by user code.

> - Adds a some amount of overhead to cross-loop communication.

Reviewer 3: nit pick: "Adds some amount" instead

                await asyncio.sleep(1)
                yield f"{i}\n"

        return generator()
```

In streaming scenarios:

- Prefer `async def` generators that use `await` between yields.
- Avoid long CPU-bound loops between yields; offload them if needed.

## Offloading patterns: I/O, CPU, and Ray tasks

This section summarizes common offloading patterns you can use inside `async` handlers.

### Blocking I/O in `async def`

```python
@serve.deployment
class OffloadIO:
    async def __call__(self, request):
        import requests

        def fetch():
            return requests.get("https://example.com/").text

        # Offload to a thread, free the event loop.
        body = await asyncio.to_thread(fetch)
        return body
```

### CPU-heavy code in `async def`

```python
@serve.deployment
class OffloadCPU:
    def _compute(self, x):
        # CPU-intensive work.
        total = 0
        for i in range(10_000_000):
            total += (i * x) % 7
        return total

    async def __call__(self, request):
        x = 123
        loop = asyncio.get_running_loop()
        result = await loop.run_in_executor(None, self._compute, x)
        return str(result)
```

> ### Using Ray tasks or remote actors for true parallelism

Author: The cluster should still be able to autoscale even if ray tasks are created from serve application. But good point about observability. Creating another deployment and offloading the work there has the same problem because then that deployment becomes blocking, so we just kick the can forward to a downstream deployment.

I have seen users use this pattern especially in OSS. So I am inclined to keep this but will mark this a advanced usage pattern, call out the observablity pitfall and not mention it repeatedly in the document.

> ### Using Ray tasks or remote actors for true parallelism

Reviewer 3: Sounds good to me about calling it an advanced pattern and leaving it to a dedicated section. 

Yes I agree remote tasks should scale just fine. In my mind with composition they can set target and max requests to a low number close to 1 and use serve's queue based replica scaling to parallelize the compute but I recognize this probably adds more performance overhead than directly submitting tasks.

As for my comment on auto scaling - I was referring to remote actors given the document originally stated "use remote tasks or actors" which would require an auto scaling actor pool abstraction...
